pkg/destroy/nutanix/nutanix.go

```diff
@@ -0,0 +1,226 @@
+package nutanix
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	nutanixclientv3 "github.com/terraform-providers/terraform-provider-nutanix/client/v3"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/openshift/installer/pkg/destroy/providers"
+	installertypes "github.com/openshift/installer/pkg/types"
+	nutanixtypes "github.com/openshift/installer/pkg/types/nutanix"
+)
+
+const (
+	emptyFilter                 = ""
+	expectedCategoryKeyFormat   = "kubernetes-io-cluster-%s"
+	expectedCategoryValueOwned  = "owned"
+	expectedCategoryValueShared = "shared"
+)
+
+// clusterUninstaller holds the various options for the cluster we want to delete.
+type clusterUninstaller struct {
+	clusterID string
+	infraID   string
+	v3Client  *nutanixclientv3.Client
+	logger    logrus.FieldLogger
+}
+
+type expectedCategory struct {
+	key    string
+	values []string
+}
+
+// New returns an Nutanix destroyer from ClusterMetadata.
+func New(logger logrus.FieldLogger, metadata *installertypes.ClusterMetadata) (providers.Destroyer, error) {
+	v3Client, err := nutanixtypes.CreateNutanixClient(context.TODO(),
+		metadata.ClusterPlatformMetadata.Nutanix.PrismCentral,
+		metadata.ClusterPlatformMetadata.Nutanix.Port,
+		metadata.ClusterPlatformMetadata.Nutanix.Username,
+		metadata.ClusterPlatformMetadata.Nutanix.Password,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &clusterUninstaller{
+		clusterID: metadata.ClusterID,
+		infraID:   metadata.InfraID,
+		v3Client:  v3Client,
+		logger:    logger,
+	}, nil
+}
+
+// Run is the entrypoint to start the uninstall process.
+func (o *clusterUninstaller) Run() (*installertypes.ClusterQuota, error) {
+	o.logger.Infof("Starting deletion of Nutanix infrastructure for Openshift cluster %s", o.infraID)
+	err := wait.PollImmediateInfinite(time.Second*30, o.destroyCluster)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to destroy cluster")
+	}
+
+	return nil, nil
+}
+
+func (o *clusterUninstaller) destroyCluster() (bool, error) {
+	cleanupFuncs := []struct {
+		name    string
+		execute func(*clusterUninstaller) error
+	}{
+		{name: "VMs", execute: cleanupVMs},
+		{name: "Images", execute: cleanupImages},
+		{name: "Categories", execute: cleanupCategories},
```

Comment:
What is the behavior if the destroyer attempts to delete the categories while there are still VMs and/or images using the categories? Will the delete fail? If the delete succeeds, will the categories be removed from the remaining VMs and images? In other words, do we need to wait until the VMs and images are deleted before attempting to delete the categories?

Reply:
No, categories can't be deleted as long as resources with the categories exist.

Reply:
Perfect. That works then. The delete of the categories will fail, which is fine.